"Prelecture" HW

In [1]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()

df.shape 

(391, 11)

Each row is an observation of the data
Each column is a variable being recorded

In [2]:
df.columns

Index(['row_n', 'id', 'name', 'gender', 'species', 'birthday', 'personality',
       'song', 'phrase', 'full_id', 'url'],
      dtype='object')

In [3]:
df.describe()

row_n
count  391.000000
mean   239.902813
std    140.702672
min      2.000000
25%    117.500000
50%    240.000000
75%    363.500000
max    483.000000

In [5]:
df.describe

<bound method NDFrame.describe of      row_n        id      name  gender    species birthday personality  \
0        2   admiral   Admiral    male       bird     1-27      cranky   
1        3   agent-s   Agent S  female   squirrel      7-2       peppy   
2        4     agnes     Agnes  female        pig     4-21        uchi   
3        6        al        Al    male    gorilla    10-18        lazy   
4        7   alfonso   Alfonso    male  alligator      6-9        lazy   
..     ...       ...       ...     ...        ...      ...         ...   
386    475    winnie    Winnie  female      horse     1-31       peppy   
387    477  wolfgang  Wolfgang    male       wolf    11-25      cranky   
388    480      yuka      Yuka  female      koala     7-20      snooty   
389    481      zell      Zell    male       deer      6-7        smug   
390    483    zucker    Zucker    male    octopus      3-8        lazy   

                song    phrase            full_id  \
0         Steep Hill   a

a) df.shape considers all columns while df.describe focuses only on the numerical columns

b) if there is a missing value in the numerical column or if the columns contain non-numerical data, then the count reported by df.describe will be lower than the count reported by df.shape, because df.describe does not include missing values and non-numerical columns.

In [ ]:
Summary of the ChatBot session:

This chat has focused on analyzing a dataset, likely related to the game Animal Crossing, using Python's pandas library. You've been working with a DataFrame (df) and have been interested in understanding its structure, contents, and potential discrepancies in the information returned by different pandas functions.
Here's a breakdown of the key points from our conversation:●
Dataset Context: While the exact source of your data isn't specified in the provided sources, your code suggests you're working with data on villagers, likely from Animal Crossing. You're using Python and the pandas library for your analysis. [You provided this code]●
Understanding DataFrames: Our conversation explored the concepts of "observations" and "variables" in the context of a dataset. We established that:○
Observations typically correspond to rows in your DataFrame (likely each row represents a villager). [Our conversation history]○
Variables typically correspond to columns in your DataFrame (characteristics like villager name, birthday, catchphrase, etc.). [Our conversation history]●
Using df.shape: You used the df.shape attribute to get the dimensions of your DataFrame (number of rows and columns). It was highlighted that df.shape considers all columns, regardless of their data type. [Our conversation history]●
Using df.describe(): You also utilized the df.describe() method, which provides descriptive statistics for numerical columns in a DataFrame. We discussed how this method, by default, ignores non-numerical columns and how missing values (NaNs) can lead to discrepancies between the counts reported by df.describe() and the total number of rows shown by df.shape. [Our conversation history]
Important Note: The provided sources primarily focus on GitHub features and navigation, offering limited direct information about pandas DataFrames or Animal Crossing data. The explanations and insights provided about analyzing your dataset were drawn from common knowledge about Python's pandas library and were not directly supported by the sources you provided.












"Postlecture" HW

6. The df.describe() method provides the 'count', 'mean', 'std', 'min', '25%', '50%', '75%', and 'max' summary statistics for each variable it analyzes. Give the definitions (perhaps using help from the ChatBot if needed) of each of these summary statistics

count: the number of data points in each column
mean: the average value of the data points in the column
std: the variation of the data from the mean
min:the smallest value in the dataset for the variable
25%:Q1, where 25% of the data falls below this point
50%:the median of the data
75%:Q3, where 75% of the data falls below this point
max:the largest value in the dataset for the variable


7. Missing data can be considered "across rows" or "down columns". Consider how df.dropna() or del df['col'] should be applied to most efficiently use the available non-missing data in your dataset and briefly answer the following questions in your own words

1.Provide an example of a "use case" in which using df.dropna() might be peferred over using del df['col']
    when you want to remove rows or columns that contains missing values, you can use df.dropna(), which automatically removes all rows or columns with at least one missing value.


2.Provide an example of "the opposite use case" in which using del df['col'] might be preferred over using df.dropna()
    when you want to remove columns that you think is irrelevant, you can choose del df['col'] to delete the irrelevant columns and get a more clean dataset


3.Discuss why applying del df['col'] before df.dropna() when both are used together could be important
    If you use del df['col'] before df.dropna(), you can delete irrelevant columns beforehand, which avoids handling unnecessary data and makes the process of df.dropna() faster and more clean.


4.Remove all missing data from one of the datasets you're considering using some combination of del df['col'] and/or df.dropna() and give a justification for your approach, including a "before and after" report of the results of your approach for your dataset.

In [2]:
import pandas as pd
url = "https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-05-05/villagers.csv"
df = pd.read_csv(url)
df.isna().sum()
df.dropna()
df.dropna(axis=1)

row_n      name  gender    species birthday personality    phrase  \
0        2   Admiral    male       bird     1-27      cranky   aye aye   
1        3   Agent S  female   squirrel      7-2       peppy  sidekick   
2        4     Agnes  female        pig     4-21        uchi   snuffle   
3        6        Al    male    gorilla    10-18        lazy   Ayyeeee   
4        7   Alfonso    male  alligator      6-9        lazy  it'sa me   
..     ...       ...     ...        ...      ...         ...       ...   
386    475    Winnie  female      horse     1-31       peppy    hay-OK   
387    477  Wolfgang    male       wolf    11-25      cranky   snarrrl   
388    480      Yuka  female      koala     7-20      snooty   tsk tsk   
389    481      Zell    male       deer      6-7        smug     pronk   
390    483    Zucker    male    octopus      3-8        lazy     bloop   

               full_id                                                url  
0     villager-admiral  https://villagerdb.com/images/villagers/thumb/...  
1     villager-agent-s  https://villagerdb.com/images/villagers/thumb/...  
2       villager-agnes  https://villagerdb.com/images/villagers/thumb/...  
3          villager-al  https://villagerdb.com/images/villagers/thumb/...  
4     villager-alfonso  https://villagerdb.com/images/villagers/thumb/...  
..                 ...                                                ...  
386    villager-winnie  https://villagerdb.com/images/villagers/thumb/...  
387  villager-wolfgang  https://villagerdb.com/images/villagers/thumb/...  
388      villager-yuka  https://villagerdb.com/images/villagers/thumb/...  
389      villager-zell  https://villagerdb.com/images/villagers/thumb/...  
390    villager-zucker  https://villagerdb.com/images/villagers/thumb/...  

[391 rows x 9 columns]

I used df.dropna() as an approach to remove all missing data from the dataset.
before applying df.dropna() to the dataset, the dataset contains 391 rows, after applying df.dropna() and removing the rows with missing values, there are 379 rows left.
I used this method because I don't see where the missing values are, so I do not have a specific column or row that I know I should remove. Therefore, I used df.dropna() and df.dropna(axis=1) to remove all rows and columns with at least 1 missing value.




Summary of Chat Session
df.describe() Method:

Count: Number of non-null observations.
Mean: Average value of the data.
Standard Deviation (std): Measure of data dispersion from the mean.
Minimum (min): Smallest value in the data.
25th Percentile (25%): Value below which 25% of the data falls.
50th Percentile (50%): Median value of the data.
75th Percentile (75%): Value below which 75% of the data falls.
Maximum (max): Largest value in the data.
df.dropna() Method:

Removes rows or columns with missing values (NaN).
Parameters include axis, how, thresh, subset, and inplace.
del df['col'] Statement:

Deletes a specific column from a DataFrame.
Modifies the DataFrame in place, does not return a new DataFrame.
Importance of Order (Deleting Columns Before Dropping NaNs):

Performance Efficiency: Reduces memory usage and speeds up operations.
Relevance: Focuses missing data handling on relevant columns.
Data Cleaning Logic: Follows a logical workflow in data cleaning.







8. Give brief explanations in your own words for any requested answers to the questions below

1.Use your ChatBot session to understand what df.groupby("col1")["col2"].describe() does and then demonstrate and explain this using a different example from the "titanic" data set other than what the ChatBot automatically provide for you

In [5]:
 import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv" 
df = pd.read_csv(url)
df.isna().sum()

df.groupby("sex")["survived"].describe()

count      mean       std  min  25%  50%  75%  max
sex                                                       
female  314.0  0.742038  0.438211  0.0  0.0  1.0  1.0  1.0
male    577.0  0.188908  0.391775  0.0  0.0  0.0  0.0  1.0

df.groupby("sex")["survived"].describe() grouped the column "sex" and then compared it to the column "survived", describing the distribution of passengers who survived between female and male.






2.Assuming you've not yet removed missing values in the manner of question "7" above, df.describe() would have different values in the count value for different data columns depending on the missingness present in the original data. Why do these capture something fundamentally different from the values in the count that result from doing something like df.groupby("col1")["col2"].describe()?
    for df.describe(), the count excludes all missing values, while the count for df.groupby("col1")["col2"].describe() exludes values based on groups. if the missing value is in one group, it only affect the count in that group, the count in other groups would not be affected.





3.Intentionally introduce the following errors into your code and report your opinion as to whether it's easier to (a) work in a ChatBot session to fix the errors, or (b) use google to search for and fix errors: first share the errors you get in the ChatBot session and see if you can work with ChatBot to troubleshoot and fix the coding errors, and then see if you think a google search for the error provides the necessary toubleshooting help more quickly than ChatGPT

A)Forget to include import pandas as pd in your code

In [1]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv" 
df = pd.read_csv(url)
df.isna().sum()

df.groupby("sex")["survived"].describe()

NameError: name 'pd' is not defined

It is easier to work in a ChatBot session to fix this error. After I share this error in the ChatBot session, it automatically gives the the code import pandas as pd and tells me to include the following import statement at the beginning of my code. When I search on google, I have to click into the websites to try to find a solution.







B)Mistype "titanic.csv" as "titanics.csv"

In [3]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanics.csv" 
df = pd.read_csv(url)
df.isna().sum()

df.groupby("sex")["survived"].describe()

HTTPError: HTTP Error 404: Not Found

I found the answer to this error rather quickly using both Google and ChatGPT, for both I had to read through some text to find the small line of answer I wanted.







C) Try to use a dataframe before it's been assigned into the variable

In [5]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv" 
df = pd.read_csv(url)
df.isna().sum()

DF.groupby("sex")["survived"].describe()

NameError: name 'DF' is not defined

ChatGPT was much more helpful for this error, it even reminds me to use df rather than DF in the summary part of its answer, whereas I got nothing on google when I enter the same thing that I asked ChatGPD in the search.








D) Forget one of the parentheses somewhere the code

In [6]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv" 
df = pd.read_csv(url)
df.isna().sum()

df.groupby("sex")["survived"].describe(

SyntaxError: incomplete input (4259902073.py, line 6)

ChatGPT was more useful, since it only gives me answers to my question. When I search on Google, the things that appeared are questions that other people asked and answered, which is not very helpful for me.







E) Mistype one of the names of the chained functions with the code

In [7]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv" 
df = pd.read_csv(url)
df.isna().sum()

df.group_by("col1")["col2"].describe()

AttributeError: 'DataFrame' object has no attribute 'group_by'

ChatGPT tells me to use groupby instead of group_by in the first three lines of the answer, while in Google I did not find much useful information from the problems that are slightly different from mine.






F) Use a column name that's not in your data for the groupby and column selection

In [8]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv" 
df = pd.read_csv(url)
df.isna().sum()

df.groupby("SEX")["survived"].describe()

KeyError: 'SEX'

The answers on Google mentioned about checking the column names, but I have to look very closely and browse through a lot of words to get a glimpse of it, whereas ChatGPT tells me to ensure that you are using the correct column name in the first line.






G) Forget to put the column name as a string in quotes for the groupby and column selection, and see if the ChatBot and google are still as helpful as they were for the previous question

In [10]:
import pandas as pd
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv" 
df = pd.read_csv(url)
df.isna().sum()

df.groupby(sex)["age"].describe()

NameError: name 'sex' is not defined

In the answer ChatGPT gives me, it says use column names as strings and ensure that they match exactly. When I search on Google, I am still browsing over informations and situations that does not match my problem.



9. Have you reviewed the course wiki-textbook and interacted with a ChatBot (or, if that wasn't sufficient, real people in the course piazza discussion board or TA office hours) to help you understand all the material in the tutorial and lecture that you didn't quite follow when you first saw it?

Mostly




Overview

We discussed various issues related to data analysis with the Titanic dataset using Python and pandas. The primary focus was on resolving errors and understanding how to use different DataFrame methods correctly.

Key Topics Covered
df.groupby() and df.describe()

df.groupby('col1')['col2'].describe(): Provides descriptive statistics for col2 within each group of col1.
df.describe(): Gives summary statistics for all numeric columns, handling missing values by default.
Handling Missing Values and Descriptive Statistics

We discussed how missing values affect the count in descriptive statistics. The count in df.describe() reflects non-missing values, while groupby with describe() provides statistics for each group based on the available data in that group.
Correct Method Usage

Correct Method: groupby (not group_by) is the method to group DataFrame data.
Case Sensitivity: Column names are case-sensitive (e.g., sex vs. SEX).
Error Resolution

NameError: name 'pd' is not defined: Ensure pandas is imported as pd.
HTTPError: HTTP Error 404: Check URL correctness or use an alternative source.
AttributeError: 'DataFrame' object has no attribute 'group_by': Use groupby instead.
KeyError: 'SEX': Verify the correct column name; it's usually 'sex'.
NameError: name 'titanic_df' is not defined: Define the variable before use.
NameError: name 'sex' is not defined: Ensure you’re using column names correctly.
Saving and Sharing Chat History

Options for saving and sharing the chat log include copying and pasting, taking screenshots, and using email or cloud services.

Example Code
Here’s an example demonstrating correct usage:

import pandas as pd

# Load the Titanic dataset from a URL
df = pd.read_csv('https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv')

# Display the first few rows
print(df.head())

# Group by 'sex' and get descriptive statistics for 'age'
grouped_by_sex = df.groupby('sex')['age'].describe()

print(grouped_by_sex)


Summary

The session focused on:

Understanding and fixing common errors related to data analysis.
Correctly using pandas methods and handling different DataFrame operations.
Ensuring proper definition and usage of variables and column names.